File: testdata/fixtures/swift/CWE-434.swift
// CWE-434: FileManager createFile without ext check
import Foundation

func vuln1(uploadedFile: (name: String, data: Data)) {
    let name = uploadedFile.name
    // ruleid: swift-434-upload
    FileManager.default.createFile(atPath: "/uploads/" + name, contents: uploadedFile.data)
}

func vuln2(file: (filename: String, content: Data)) {
    let path = "/tmp/" + file.filename
    // ruleid: swift-434-upload
    FileManager.default.createFile(atPath: path, contents: file.content)
}

func vuln3(requestData: (fileName: String, bytes: Data)) {
    let p = requestData.fileName
    // ruleid: swift-434-upload
    Data.write(to: URL(fileURLWithPath: p))
}

func safeAllowlist(uploadedFile: (name: String, data: Data)) {
    let ext = (uploadedFile.name as NSString).pathExtension
    if !["png","jpg"].contains(ext) { return }
    // ok: swift-434-upload
    FileManager.default.createFile(atPath: "/uploads/allowed.png", contents: uploadedFile.data)
}

func safeFixed() {
    // ok: swift-434-upload
    FileManager.default.createFile(atPath: "/tmp/fixed.png", contents: Data())
}

File: testdata/fixtures/python/CWE-434.py
# CWE-434: Unrestricted upload — filename from request.files reaches file write without whitelist.
from flask import request
import os

def upload_vuln():
    f = request.files.get('file')
    name = f.filename
    # ruleid: python-unrestricted-upload
    open(os.path.join('/uploads', name), 'wb').write(f.read())
    # ruleid: python-unrestricted-upload
    f.save(os.path.join('/tmp', name))

def upload_vuln_s3():
    uploaded = request.files['file']
    fname = uploaded.filename
    # ruleid: python-unrestricted-upload
    s3.upload_file(os.path.join('/tmp', fname), 'mybucket', fname)

def safe_fixed():
    # ok: python-unrestricted-upload
    open('/tmp/fixed.txt', 'wb').write(b'fixed')
    # ok: python-unrestricted-upload
    s3.upload_file('/tmp/fixed.txt', 'mybucket', 'fixed.txt')

def safe_whitelist():
    f = request.files.get('file')
    name = f.filename
    if not name.endswith('.png'):
        raise ValueError('bad extension')
    # ok: python-unrestricted-upload
    open(os.path.join('/uploads', 'safe.png'), 'wb').write(f.read())
